### ISSUE TITLE
QMMM + Solvent

### ISSUE THREAD

**Issue by swillow** [pyscf/pyscf#446]
I wonder whether QM/MM + Solvent is supported.
If not, I wish that pyscf supports the QM/MM + Solvent.

**Comment by sunqm** [comment]
Technically, the program supports to apply the implicit solvent model and QM/MM charges in the same calculation. E.g.
https://github.com/pyscf/pyscf/blob/master/examples/solvent/22-with_qmmm.py

However, the cavity of solvent in current implementation does not count the MM "particles". It is placed around the QM region. I'm not sure whether this is the true QM/MM + solvent .

**Comment by swillow** [comment]
The following is my understanding of the QM/MM/solvent system.
----
In the QM/solvent, the cavity of solvent is the QM region.
However, in the QM/MM/solvent, the cavity of solvent should be the QM/MM region.
In addition, we need to consider the interaction between MM (charges) and solvent as the interaction between QM (\rho_electron + {Z}) and solvent.

**Comment by swillow** [comment]
I ask the cavity of the solvent for the QM/MM + solvent system.

The following is the e-mail from Prof. Filippo Lipparini.

'For QM/MM calculations, we use a cavity that accommodates the overall
system - both QM and MM atoms are thus considered when building the
cavity. In particular, as we use a (scaled) Van der Waals cavity, in our
implementation we assume that there is a sphere placed on each atom -
without distinction on whether it is a QM or MM one: the cavity is then
obtained as the union of all the spheres. Gridpoints are placed on the
surface of every sphere, which is what makes a QM/MM/Continuum
calculation very expensive. However, ddCOSMO should easily manage up to
a few tens of thousands atoms.

One suggestion: Van der Waals cavity do an excellent job with small
molecules, however, when used on very large systems - such as proteins -
it is very important to have a look at the surface that you get, as
there might be holes, channels and cavities that allow the solvent to
enter physically inaccessible regions. For this reason, I believe that
it is a better idea to use a solvent accessible surface, which you can
easily get by using Van der Waals radii incremented by the effective
radius of the solvent that you are using (1.4 Angstrom for water). This
gives usually more robust results.'

**Comment by sunqm** [comment]
The combination QM/MM + implicit solvent was considered in the latest update. Please the example https://github.com/pyscf/pyscf/blob/master/examples/solvent/22-with_qmmm.py
In this example, the MM energy and the vdw/non-bonding interaction between MM and QM systems were not included.

**Comment by swillow** [comment]
I have one minor comment.
The following line 
'sol.radii_table = radii.VDW'
can be improved into
'sol_radius = 1.4 # A (The length unit should be careful, here)
sol.radii_table = radii.VDW + sol_radius'
Here, sol_radius will be a user-defined solvent radius (default value is 1.4 A)

It is necessary to mention that the Carbon is used for the element of MM charges. 
Hence, the user should provide correct elements in class QMMMMole.

Besides, the interactions between MM and QM systems can be handled by the other packages (OpenMM, Amber, Charmm, ...) once the VDW parameters and RESP atomic charges for the QM molecule are provided (in my case, I used the ambertools to generate the VDW parameters).

Thank you for your help and providing an excellent example quickly.

**Comment by swillow** [comment]
Dear Qiming

I tried to run a QMMM/Solvent with a ligand-protein cmplex.
Here, ligand (49 atoms) belongs to the QM and the protein (3139 atoms)  is in the MM region.

I could obtain the total potential energy of the ligand with the ddCOSMO solvent.
I could get the total potential energy of the ligand-complex using the QM/MM scheme.
However, I could not obtain the total potential energy of the ligand-complex with the ddCOSMO solvent.

Firstly, I realized that the grids of QMMM atoms (49+3139 atoms) are built:
        https://github.com/pyscf/pyscf/blob/8af0e4a69374508c9ebd4dbdff404bd8976bb5d2/pyscf/solvent/ddcosmo.py#L1044

       https://github.com/pyscf/pyscf/blob/8af0e4a69374508c9ebd4dbdff404bd8976bb5d2/pyscf/solvent/ddcosmo.py#L702

This becomes the rate determinant step.

To avoid this problem, I modified 'class DDCOSMO' 
---
def __init__ (self, mol, mol_qm = None):
...(line 1044)
if mol_qm is not None:
     self.grids = gen_grid.Grids(mol_qm)
else:
    self.grids = gen_grid.Grids(mol)
--

Even though I modified the above lines, I still have other problems: 
https://github.com/pyscf/pyscf/blob/8af0e4a69374508c9ebd4dbdff404bd8976bb5d2/pyscf/solvent/ddcosmo.py#L815
I got a Memory Error in Lmat owing to the huge number of atoms in QMMM.

In short, my question is 
1) Is it OK to build the grids of the QM zone for \rho(R)? (or we have to generate all grids of the QMMM zone.)
2) I thought simply that L X = g for the QM zone with the cavity of the QMMM zone. The solvent boundary is based on the QMMM zone and the other calculations are the same the QM zone.
Is it possible to approach this above scheme?

Thank you for your help in advance.

**Comment by swillow** [comment]
To be clear, let us start with Eqs. (26) ~ (28) in JCP 141, 184108 (2014).
In case of the QM/ddCOSMO, the numerical quadrature procedure was implemented to replace the integral over \rho(r).

In case of the QMMM/ddCOSMO, we have the charge density of \rho_QM(r) in the QM region and the \rho_MM(R) = \sum_{j \in MM} q_j \delta (R-R_j).
Hence, I think that we don't need to build the grids of the QM/MM region.
Based on Eq. (31), we don't need to use grid point \rho(x_n^j) for the MM region. It will be much simpler even though I don't have a proper mathematical expression.
